### PR TITLE
Handle 'nul' or 'NUL' devices on Windows platform.

### PR DESCRIPTION
**Pull request by weitjong**
This is useful in the following use case: `emcc.bat -E -dM -xc nul`

In our cross-platform project we use CMake to execute a process to dump out a list of 'native' compiler predefined macros. In Unix-alike platforms, the generated command looks like this:

```
emcc -E -dM -xc /dev/null
```

and it works as expected. However, on Windows platform current `emcc` does not understand `nul` as null device. This PR fixes this issue.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -34,10 +34,13 @@
 CXX_ENDINGS = ('.cpp', '.cxx', '.cc', '.c++', '.CPP', '.CXX', '.CC', '.C++', '.ii')
 OBJC_ENDINGS = ('.m', '.mi')
 OBJCXX_ENDINGS = ('.mm', '.mii')
-SPECIAL_ENDINGLESS_FILENAMES = ('/dev/null',)
+SPECIAL_ENDINGLESS_FILENAMES = {
```
Comment (kripken) [inline emcc.py]: a command explaining what True and False mean here would be nice

**Comment by kripken** [comment]
Overall looks ok to me, but I don't know much about windows. @juj?

We should add a test for this, in `tests/test_other.py`.

**Comment by weitjong** [comment]
Thanks for reviewing the code. I have put a comment and a simple test case in as per your request.

File: tests/test_other.py [hunk 2]
```diff
@@ -3595,6 +3595,10 @@ def test_dashM_respect_dashO(self):
     assert len(with_dash_o) == 0
     assert len(without_dash_o) != 0
 
+  def test_null_device_as_input(self):
+    out = Popen([PYTHON, EMXX, '-E', '-dM', '-xc', '/dev/null'], stdout=PIPE).communicate()[0]
```
Comment (kripken) [inline tests/test_other.py]: I think the test should do something different on windows than on other platforms?

File: tests/test_other.py [hunk 3]
```diff
@@ -3595,6 +3595,10 @@ def test_dashM_respect_dashO(self):
...
+    self.assertContained('#define __EMSCRIPTEN__ 1', out) # Verify output contains one of the native predefined macros
```
Comment (kripken) [inline tests/test_other.py]: this comment still doesn't make it clear enough to me why checking for that output is enough for this test. When that output is present, what does it prove worked?

File: tests/test_other.py (hunk 2 above)
Comment (weitjong) [inline tests/test_other.py]: You are right, however, I have no idea how the Emscripten's unit testing is being carried out on Windows. The test scripts appear to be *nix-centric to me. At the very least, this test ensures that the changes in this PR does not cause any regression issue on *nix platforms.

If the test suite could be run on Windows then we could add another similar test as above but for Windows by simply replacing "/dev/null" to "nul".

File: tests/test_other.py (hunk 3 above)
Comment (weitjong) [inline tests/test_other.py]: If the null device file was not handled correctly by emcc then it would not produce anything in the stdout, but just error complaining about missing input file (or something like that) in the stderr. So, actually by checking there is some output in stdout and nothing in stderr are sufficient. Here, I went a little bit further to check the output is conforming to "-E -dM" compiler flags. Let me know if you want me to just check "there is output in stdout and nothing in stderr" instead.

File: tests/test_other.py (hunk 2 above)
Comment (kripken) [inline tests/test_other.py]: Yes, I think we can just do `"/dev/null" if not WINDOWS else "nul"` in this test, and then it would run on all platforms.

The test suite is run on windows machine on our bots, however some tests might be disabled on windows, so perhaps you saw one of those? @juj, do we not run all `other` tests on windows?

File: tests/test_other.py (hunk 3 above)
Comment (kripken) [inline tests/test_other.py]: In that case, you can just do `subprocess.check_call` and it checks the return code, verifying that it ran without hitting an error (it would have failed before, I believe?).

**Comment by weitjong** [comment]
While about to make the changes as per your recommendation, I spot an existing unit test that performs almost exactly as what I have proposed. So, I push another commit to drop my new unit test and instead amend the existing one so it could work on all platforms as you want. The test now checks on return code of the subprocess, and also stdout and stderr.

File: emcc.py [hunk 4]
```diff
@@ -34,10 +34,15 @@
 CXX_ENDINGS = ('.cpp', '.cxx', '.cc', '.c++', '.CPP', '.CXX', '.CC', '.C++', '.ii')
 OBJC_ENDINGS = ('.m', '.mi')
 OBJCXX_ENDINGS = ('.mm', '.mii')
-SPECIAL_ENDINGLESS_FILENAMES = ('/dev/null',)
+# A map of platform-specific 'null device' file(s)
+# There are only two keys in the map: False for non-Windows (or for POSIX to be more precise) and True for Windows
+SPECIAL_ENDINGLESS_FILENAMES = {
+  False: ('/dev/null',),
+  True: ('nul', 'NUL')
+}
```
Comment (juj) [inline emcc.py]: Feels like indexing with a boolean like this is a bit non-idiomatic and down below with the `SPECIAL_ENDINGLESS_FILENAMES[WINDOWS]` as well. How about the code would look like

    SPECIAL_ENDINGLESS_FILENAMES = ['nul', 'NUL'] if WINDOWS else ['/dev/null']

then any other locations that use the value don't need to specifically index with `[WINDOWS]`, and it would look cleaner and not need a special comment to explain. Likewise, I'd prefer using a list here rather than a tuple, since a list feels like it makes more sense.

On Windows, I think we could as well support `/dev/null`, because it would then just work for MSYS/Cygwin, and I don't think anyone is looking to have a file `C:/dev/null` with meaningful content anyways. That might help some random build script emulate support on Windows, with probably no drawbacks. So we could have this even as

    SPECIAL_ENDINGLESS_FILENAMES = ['/dev/null']
    if WINDOWS: SPECIAL_ENDINGLESS_FILENAMES += ['nul', 'NUL']

Also, I think we could just rename `SPECIAL_ENDINGLESS_FILENAMES` to `FILENAMES_REPRESENTING_NULL_FILE` since there's only files representing null there (if we do later have some more in there, we could just rename or add some other name).

@kripken: would you agree with this?

Comment (kripken) [inline emcc.py]: Yeah, that sounds nice.

**Comment by kripken** [comment]
This PR is quite old, closing. Feel free to reopen if still relevant.
